### PR TITLE
feat: CSP enhancements

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
- Auto  CSP tag replacement
  - add CSP tag customization for security
  - add config for disabling auto CSP tag replacement
- add `csp_style_nonce()` and `csp_script_nonce()` to get nonce attributes
- add `csp_style_nonce` and `csp_script_nonce` plugin for View Parser
- output one nonce for script and one for style tags
  - multiple nonces do not improve security
- add `Services::csp()`

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Common.php [hunk 1]
```diff
@@ -21,6 +21,7 @@
 use CodeIgniter\HTTP\Exceptions\HTTPException;
 use CodeIgniter\HTTP\RedirectResponse;
 use CodeIgniter\HTTP\RequestInterface;
+use CodeIgniter\HTTP\Response;
```
Comment (MGatner) [inline system/Common.php:24]: I know our HTTP interfaces and instances are a total mess. But why was this added? I don't see it in your new methods.

File: system/Config/Services.php [hunk 2]
```diff
@@ -153,6 +155,22 @@ public static function commands(bool $getShared = true)
         return new Commands();
     }
 
+    /**
+     * Content Security Policy
+     *
+     * @return ContentSecurityPolicy
+     */
+    public static function contentsecuritypolicy(?CSPConfig $config = null, bool $getShared = true)
```
Comment (MGatner) [inline system/Config/Services.php:163]: Would `csp()` be preferable? Or is that too obscure?

File: system/Debug/Kint/RichRenderer.php [hunk 3]
```diff
@@ -36,11 +36,11 @@ public function preRender()
 
                 switch ($type) {
                     case 'script':
-                        $output .= '<script {csp-script-nonce} class="kint-rich-script">' . $contents . '</script>';
+                        $output .= '<script ' . csp_script_nonce() . ' class="kint-rich-script">' . $contents . '</script>';
```
Comment (MGatner) [inline system/Debug/Kint/RichRenderer.php:39]: Will these work with View Parser? I don't use it myself but I know people like their Twig-like variable injection...

File: tests/system/CommonFunctionsTest.php [hunk 4]
```diff
@@ -486,19 +486,20 @@ public function testIsCli()
 
     public function testDWithCSP()
     {
+        $this->resetServices();
+        $this->resetFactories();
```
Comment (MGatner) [inline tests/system/CommonFunctionsTest.php:490]: Factories are already reset by default during `CIUnitTestCase::setUp()` via `$setUpMethods`.

**Review by MGatner — CHANGES_REQUESTED** [review]
This looks like a great PR! Per our recent semver conversation: would this be best held for 4.2? I know the Kint changes were already merged...

I added a few questions. I'm abashedly not very up on CSP. I will read up on it a bit but would also appreciate if someone with better security mindset took a look. @lonnieezell maybe?

File: system/Config/Services.php (hunk 2 above)
Comment (lonnieezell) [reply system/Config/Services.php:163]: This not being camel case makes me twitch. :) I think `csp` is fine, since the helper methods are `csp_x`.

**Review by lonnieezell — COMMENTED** [review]
I like these changes a lot and don't see any security concerns with it.

File: system/Common.php (hunk 1 above)
Comment (kenjis) [reply system/Common.php]: Good catch. It is because CSP instance was created in the Response at first.

Removed.

File: system/Config/Services.php (hunk 2 above)
Comment (kenjis) [reply system/Config/Services.php]: Changed.

File: tests/system/CommonFunctionsTest.php (hunk 4 above)
Comment (kenjis) [reply tests/system/CommonFunctionsTest.php]: This seems to be my mistake. It is not needed.

Removed.

File: system/Debug/Kint/RichRenderer.php (hunk 3 above)
Comment (kenjis) [reply system/Debug/Kint/RichRenderer.php:39]: Good point. I added plugins for View Parser.

**Review by MGatner — APPROVED** [review]
Very impressive test-fu.

**Comment by kenjis** [comment]
I will add the documentation.

**Comment by kenjis** [comment]
Added the documentation. Review, please.

File: tests/system/CommonFunctionsTest.php [hunk 5]
```diff
@@ -541,12 +541,34 @@ public function testDWithCSP()
...
     }
+
+    public function testCspStyleNonce()
+    {
+        $this->resetServices();
+
+        $config             = config('App');
+        $config->CSPEnabled = true;
+
+        $this->assertStringContainsString('nonce="', csp_style_nonce());
```
Comment (paulbalandan) [inline tests/system/CommonFunctionsTest.php:562]: Can you change to `assertStringStartsWith`?

File: tests/system/CommonFunctionsTest.php [hunk 6]
```diff
@@ -541,12 +541,34 @@ public function testDWithCSP()
...
+    }
+
+    public function testCspScriptNonce()
+    {
+        $this->resetServices();
+
+        $config             = config('App');
+        $config->CSPEnabled = true;
+
+        $this->assertStringContainsString('nonce="', csp_script_nonce());
```
Comment (paulbalandan) [inline tests/system/CommonFunctionsTest.php:572]: Same here

File: user_guide_src/source/general/common_functions.rst [hunk 7]
```diff
@@ -217,6 +217,22 @@ Miscellaneous Functions
 
     Returns the timezone the application has been set to display dates in.
 
+.. php:function:: csp_script_nonce ()
+
+    :returns: The CSP nonce attribute for script tag.
+    :rtype: string
+
+    Returns the nonce attribute for script tag. That is string like ``nonce="Eskdikejidojdk978Ad8jf"``.
```
Comment (MGatner) [inline user_guide_src/source/general/common_functions.rst:225]:
```suggestion
    Returns the nonce attribute for a script tag. For example: ``nonce="Eskdikejidojdk978Ad8jf"``.
```

File: user_guide_src/source/general/common_functions.rst [hunk 8]
```diff
@@ -217,6 +217,22 @@ Miscellaneous Functions
...
+    See :ref:`content-security-policy`.
+
+.. php:function:: csp_style_nonce ()
+
+    :returns: The CSP nonce attribute for style tag.
+    :rtype: string
+
+    Returns the nonce attribute for style tag. That is string like ``nonce="Eskdikejidojdk978Ad8jf"``.
```
Comment (MGatner) [inline user_guide_src/source/general/common_functions.rst:233]:
```suggestion
    Returns the nonce attribute for a style tag. For example: ``nonce="Eskdikejidojdk978Ad8jf"``.
```

File: user_guide_src/source/outgoing/response.rst [hunk 9]
```diff
@@ -249,6 +251,27 @@ life, and is most secure when generated on the fly. To make this simple, you can
         . . .
     </style>
 
+.. warning:: If an attacker injects string like ``<script {csp-script-nonce}>``, it might become the real nonce attribute with this functionality. You can customize the placeholder string with the ``$scriptNonceTag`` and ``$styleNonceTag`` properties in **app/Config/ContentSecurityPolicy.php**.
```
Comment (MGatner) [inline user_guide_src/source/outgoing/response.rst:254]:
```suggestion
.. warning:: If an attacker injects a string like ``<script {csp-script-nonce}>``, it might become the real nonce attribute with this functionality. You can customize the placeholder string with the ``$scriptNonceTag`` and ``$styleNonceTag`` properties in **app/Config/ContentSecurityPolicy.php**.
```

**Review by MGatner — CHANGES_REQUESTED** [review]
Really good examples in the docs. A couple wording suggestions. Also, please mention this as a new feature in the upgrade guides so we don't new to go back and add it.

File: tests/system/CommonFunctionsTest.php (hunk 5 above)
Comment (kenjis) [reply tests/system/CommonFunctionsTest.php]: Thanks! Fixed.

File: tests/system/CommonFunctionsTest.php (hunk 6 above)
Comment (kenjis) [reply tests/system/CommonFunctionsTest.php]: Fixed.

**Comment by kenjis** [comment]
> Also, please mention this as a new feature in the changelog/upgrade guides so we don't need to go back and add it.

Added changelogs/v4.2.0.

**Review by paulbalandan — APPROVED** [review]

File: user_guide_src/source/changelogs/v4.2.0.rst [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+Changes
+*******
+
+- The current version of Content Security Policy (CSP) outputs one nonce for script and one for style tags. The previous version outputted one nonce for each tag.
+
+Deprecations
+************
+
+
```
Comment (MGatner) [inline user_guide_src/source/changelogs/v4.2.0.rst:34]:
```suggestion
* `Response`/`ResponseTrait` property `$CSPEnabled` is deprecated; use `$this->CSP->enabled()` instead.
```

**Review by MGatner — CHANGES_REQUESTED** [review]

**Review by lonnieezell — APPROVED** [review]

**Review by MGatner — APPROVED** [review]

**Comment by MGatner** [comment]
Let's hold this for 4.2
